stan/math/rev/core/profiling.hpp

```diff
@@ -0,0 +1,179 @@
+ * with the specified name and adds a var that stops
+ * profiling the reverse pass in the reverse callback.
+ * 
+ * @tparam T profile type which must not a var
+ * 
+ * @param name the name of the profile to start
+ * @param profiles the map used for storing profiling info
+ */
+template <typename T, require_all_var_t<T>* = nullptr>
+inline auto profile_start(const char* name, profile_info& profile) {
+  using std::chrono::duration;
+  using std::chrono::steady_clock;
+  if (profile.meta.fwd_pass_active) {
+    std::ostringstream msg;
+    msg << "Profile '" << name << "' ";
+    msg << "already started!";
+    throw std::runtime_error(msg.str());
+  }
+  profile.meta.fwd_pass_start_tp = steady_clock::now();
+  profile.meta.fwd_pass_active = true;
+  profile.meta.start_chain_stack_size = ChainableStack::instance_->var_stack_.size();
+  reverse_pass_callback([name, &profile]() mutable {
+    profile.rev_pass_time += duration<double>(steady_clock::now()
+                             - profile.meta.rev_pass_start_tp).count();
+    profile.meta.rev_pass_active = false;
+    profile.n_rev_pass++;
+  });
+}
+
+/**
+ * Stops profiling the forward pass for the profile
+ * with the specified name and adds a var that starts
+ * profiling the reverse pass in the reverse callback.
+ * 
+ * @tparam T profile type which must not a var
+ * 
+ * @param name the name of the profile to stop
+ * @param profiles the map used for storing profiling info
+ */
+template <typename T, require_all_var_t<T>* = nullptr>
+inline auto profile_stop(const char* name, profile_info& profile) {
+  using std::chrono::duration;
+  using std::chrono::steady_clock;
+  profile.fwd_pass_time += duration<double>(steady_clock::now()
+                           - profile.meta.fwd_pass_start_tp).count();
+  profile.meta.fwd_pass_active = false;
+  profile.n_fwd_pass++;
+  profile.chain_stack_size_sum += 
+    (ChainableStack::instance_->var_stack_.size() - profile.meta.start_chain_stack_size - 1);  
+  profile.nochain_stack_size_sum += 
+    (ChainableStack::instance_->var_nochain_stack_.size() - profile.meta.start_nochain_stack_size - 1);
+  reverse_pass_callback([name, &profile]() mutable {
+    profile.meta.rev_pass_start_tp = steady_clock::now();
+    profile.meta.rev_pass_active = true;
+  });
+}
+}  // namespace internal
+
+/**
+ * Profiles C++ lines where the object is in scope.
+ * When T is var, the constructor starts the profile for the forward pass
+ * and places a var with a callback to stop the reverse pass on the
+ * AD tape. The destructor stops the profile for the forward pass and
+ * places a var with a callback to start the profile for the reverse pass.
+ * When T is not var, the constructor and destructor only profile the
+ * 
+ * 
+ * @tparam T type of profile class. If var, the created object is used
+ * to profile reverse mode AD. Only profiles the forward pass otherwise.
+ */ 
+template <typename T>
+class profile {
+  const char* name_;
```

Comment:
Honestly now that I made you change these I'm scared that the compiler might put the strings at different locations. Cuz the map here is going to work on the pointer value, not the string value.

Reply:
Gladly move back to strings :)